### PR TITLE
Jaicp Livechats, channel invocation api, requestContext in BotContextManager

### PR DESCRIPTION
**Pull request by Denire**
This PR adds support for livechat channels with JAICP. After this pr is merged (plus some changes in JAICP), use will be able to create an operator channel and connect it to JAICF bot channel.

**How it's implemented:**

1. All channels who implement `JaicpCompatibleAsyncReactions` will now have method `switchToOperator`. Currently, these channels are: `ChatWidgetChannel`, `FacebookChannel`, `SlackChannel` and `TelegramChannel`. After `switchToOperator` is invoked, JAICP ChatAdapter will act as a gateway between operator and client. After livechat is finished (by user or operator) JAICF bot will receive an event livechatFinished via `BotGateway`.
2. `BotGateway` - a new entity allowing channels to process requests from external source. The method `processGatewayRequest` is used to process livechat events, but in general it allows to send events or queries to bot using client identifier.

**Other changes:**

1. `BotContextManager` now receives `RequestContext` in `loadContext` and `saveContext` methods for user to have more flexibility on context persistence management.
2. `JaicpBotResponse` now contains missing requestData field, which caused minor bugs.

### CODE REVIEW

File: core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.*
+import kotlin.random.Random
+
+/**
+ * Base class for all channels able to process requests from external service.
+ * Allows to send [InvocationEventRequest] or [InvocationQueryRequest] with client identifier to implementations at any time.
+ *
+ * @see InvocationRequest
+ *
+ * */
```
Comment (morfeusys) [inline core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt:16]: I do not understand this comment. What actually does this interface's implementations?

File: core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+interface InvocableBotChannel : WithLogger {
+    /**
+     * Processes an [InvocationRequest]
+     *
+     * @param request an [InvocationRequest]
+     * @param requestContext additional general request's data that can be used during the request processing
+     *
+     * @see InvocableBotChannel
+     */
+    fun processExternalInvocation(request: InvocationRequest, requestContext: RequestContext)
```
Comment (morfeusys) [inline core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt:26]: Why not `processInvocation`?

File: core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    /**
+     * Provides a messageId for substitution in request template
+     * */
+    fun provideMessageId(): String = UUID.randomUUID().toString()
+
+    /**
+     * Provides a timestamp for substitution in request template
+     * */
+    fun provideTimestamp(): String = System.currentTimeMillis().toString()
```
Comment (morfeusys) [inline core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt:36]: Why this and function above should be a part of the interface? It's template-specific data that should be calculated during every template rendering process, isn't it?

File: core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+     *
+     * @return serialized JSON with substituted request parameters
+     * */
+    @Suppress("RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+    fun getRequestTemplateFromResources(request: InvocationRequest, resourceName: String) =
+        this.javaClass.getResource(resourceName).readText()
+            .replace("{{ clientId }}", request.clientId)
+            .replace("{{ text }}", request.input)
+
+            .replace("\"{{ timestamp }}\"", provideTimestamp())
```
Comment (morfeusys) [inline core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt:49]: Why do you place a channel-specific fields to the common template?

File: core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocationQueryParams.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        else -> error("event or query must be specified in query parameters")
+    }
+
+    val input = when (type) {
+        InvocationRequestType.EVENT -> requireNotNull(event)
+        InvocationRequestType.QUERY -> requireNotNull(query)
+    }
+
+    val clientId: String = requireNotNull(queryParamsMap["clientId"]?.firstOrNull()) {
+        "clientId path variable must be specified for gateway call"
```
Comment (morfeusys) [inline core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocationQueryParams.kt:36]: An old comment obviously?

File: core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocationRequest.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package com.justai.jaicf.channel.invocationapi
+
+import com.justai.jaicf.api.BotRequest
+import com.justai.jaicf.api.EventBotRequest
+import com.justai.jaicf.api.QueryBotRequest
+
+val BotRequest.invocationApi get() = this as? InvocationRequest
```
Comment (morfeusys) [inline core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocationRequest.kt:7]: `invocationApi` is more suitable for some sort of API, not an data class that contains a request's details.

File: channels/jaicp/src/main/kotlin/com/justai/jaicf/channel/jaicp/JaicpEvents.kt [hunk 7]
```diff
@@ -3,11 +3,17 @@ package com.justai.jaicf.channel.jaicp
...
+ * @property noLivechatOperatorsOnline - is sent when scenario attempted switchToLiveChat, but no operators were online.
  *
  * @see com.justai.jaicf.channel.jaicp.channels.TelephonyEvents
+ * @see com.justai.jaicf.channel.jaicp.reactions.switchToLiveChat
  * */
+@Suppress("MemberVisibilityCanBePrivate")
 object JaicpEvents {
     const val fileEvent = "fileEvent"
+    const val liveChatFinished = "livechatFinished"
+    const val noLivechatOperatorsOnline = "noLivechatOperatorsOnline"
```
Comment (morfeusys) [inline channels/jaicp/src/main/kotlin/com/justai/jaicf/channel/jaicp/JaicpEvents.kt:18]: Maybe it's better to use upper-case format?

File: channels/jaicp/src/main/kotlin/com/justai/jaicf/channel/jaicp/http/ChatAdapterConnector.kt [hunk 8]
```diff
@@ -42,4 +45,40 @@ internal class ChatAdapterConnector(
             logger.debug("Failed to produce logs to JAICP", e)
         }
     }
+
+    fun initLiveChat(liveChatInitRequest: LiveChatInitRequest) {
+        try {
+            initLiveChat0(liveChatInitRequest)
```
Comment (morfeusys) [inline channels/jaicp/src/main/kotlin/com/justai/jaicf/channel/jaicp/http/ChatAdapterConnector.kt:51]: Why not to inline? Looks like `initLiveChat0` is not used from anywhere more?

File: channels/jaicp/src/main/kotlin/com/justai/jaicf/channel/jaicp/reactions/JaicpAsyncReactions.kt [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+ * */
+fun JaicpCompatibleAsyncReactions.switchToLiveChat(reply: LiveChatSwitchReply): SwitchReaction? {
+    val switchRequest = LiveChatInitRequest.create(loggingContext, reply) ?: return null
+    val connector = ChatAdapterConnector.getIfExists() ?: return null
+    connector.initLiveChat(switchRequest)
+    return SwitchReaction.fromReply(switchRequest.switchData, loggingContext.botContext.dialogContext.currentState)
+        .also { loggingContext.reactions.add(it) }
+}
+
+data class SwitchReaction(
```
Comment (morfeusys) [inline channels/jaicp/src/main/kotlin/com/justai/jaicf/channel/jaicp/reactions/JaicpAsyncReactions.kt:42]: I'd move this class to the upper level from this interface's file

**Review by morfeusys — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt (hunk 1 above)
Comment (Denire) [reply core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocableBotChannel.kt:16]: Any class implementing `InvocableBotChannel` receives method `processInvocation` allowing to process query or event requests with client identifier from any external source.

File: core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocationRequest.kt (hunk 6 above)
Comment (Denire) [reply core/src/main/kotlin/com/justai/jaicf/channel/invocationapi/InvocationRequest.kt]: Renamed to `invocationRequest`

**Review by morfeusys — APPROVED** [review]

**Pull request merged**